diff --git a/libtcc.c b/libtcc.c
index 10e65fb..602f849 100644
--- a/libtcc.c
+++ b/libtcc.c
@@ -721,6 +721,7 @@ PUB_FUNC void _tcc_error(const char *fmt, ...)
     va_start(ap, fmt);
     error1(ERROR_ERROR, fmt, ap);
     exit(1);
+    __builtin_unreachable();
 }
 #define _tcc_error use_tcc_error_noabort
 
diff --git a/tcc.h b/tcc.h
index 2884259..448df52 100644
--- a/tcc.h
+++ b/tcc.h
@@ -183,7 +183,8 @@
 #endif
 
 /* only native compiler supports -run */
-#if defined _WIN32 == defined TCC_TARGET_PE \
+#if !defined(TCC_TARGET_SMALLOS) \
+    && defined _WIN32 == defined TCC_TARGET_PE \
     && defined __APPLE__ == defined TCC_TARGET_MACHO
 # if defined __i386__ && defined TCC_TARGET_I386
 #  define TCC_IS_NATIVE
@@ -311,6 +312,10 @@
 #  define CONFIG_TCC_ELFINTERP "/lib/ld-linux.so.3"
 #  define CONFIG_TCC_ELFINTERP_ARMHF "/lib/ld-linux-armhf.so.3"
 # else
+#  if defined(TCC_TARGET_SMALLOS)
+#   define CONFIG_TCC_ELFINTERP "/lib/ld-smallos.so"
+#  else
 #  define CONFIG_TCC_ELFINTERP "/lib/ld-linux.so.2"
+#  endif
 # endif
 #endif
diff --git a/tccelf.c b/tccelf.c
index 0950365..b2efb0d 100644
--- a/tccelf.c
+++ b/tccelf.c
@@ -1765,6 +1765,10 @@ ST_FUNC void tccelf_add_crtbegin(TCCState *s1)
         tcc_add_crt(s1, "crtbeginS.o");
     else
         tcc_add_crt(s1, "crtbegin.o");
+#elif defined(TCC_TARGET_SMALLOS)
+    if (s1->output_type != TCC_OUTPUT_DLL)
+        tcc_add_crt(s1, "crt0.o");
+    s1->nostdlib_paths = 0;
 #elif TARGETOS_FreeBSD || TARGETOS_NetBSD
     if (s1->output_type != TCC_OUTPUT_DLL)
 #if TARGETOS_FreeBSD
@@ -1798,6 +1802,8 @@ ST_FUNC void tccelf_add_crtend(TCCState *s1)
         tcc_add_crt(s1, "crtendS.o");
     else
         tcc_add_crt(s1, "crtend.o");
+#elif defined(TCC_TARGET_SMALLOS)
+    (void)s1;
 #elif TARGETOS_FreeBSD || TARGETOS_NetBSD
     if (s1->output_type == TCC_OUTPUT_DLL)
         tcc_add_crt(s1, "crtendS.o");
@@ -1851,6 +1857,12 @@ ST_FUNC void tcc_add_runtime(TCCState *s1)
 #endif
         if (lpthread)
             tcc_add_library(s1, "pthread");
+#ifdef TCC_TARGET_SMALLOS
+        tcc_add_library(s1, "c");
+        tcc_add_library(s1, "posix");
+        tcc_add_library(s1, "c");
+        tcc_add_library(s1, "m");
+#else
         tcc_add_library(s1, "c");
 #ifdef TCC_LIBGCC
         if (!s1->static_link) {
@@ -1865,6 +1877,7 @@ ST_FUNC void tcc_add_runtime(TCCState *s1)
 #endif
         if (TCC_LIBTCC1[0])
             tcc_add_support(s1, TCC_LIBTCC1);
+#endif
 #ifndef TCC_TARGET_MACHO
         if (s1->output_type != TCC_OUTPUT_MEMORY)
             tccelf_add_crtend(s1);
